src/server/mod.rs

```diff
@@ -0,0 +1,191 @@
+use async_nats::{Client, Subscriber};
+use futures::StreamExt;
+use tracing::{info, instrument, warn};
+
+use crate::storage::Store;
+
+mod handlers;
+mod parser;
+mod types;
+
+use handlers::Handler;
+pub use parser::CONTENT_TYPE_HEADER;
+pub use types::*;
+
+/// The default topic prefix for the wadm API;
+pub const DEFAULT_WADM_TOPIC_PREFIX: &str = "wadm.api";
+
+const QUEUE_GROUP: &str = "wadm_server";
+
+/// A server for the wadm API
+pub struct Server<S> {
+    handler: Handler<S>,
+    subscriber: Subscriber,
+    prefix: String,
+}
+
+impl<S: Store + Send + Sync> Server<S> {
+    /// Returns a new server configured with the given store, NATS client, and optional topic
+    /// prefix. Returns an error if it can't subscribe on the right topics
+    ///
+    /// In most cases, you shouldn't need a custom topic prefix, but it is exposed for the cases
+    /// when you may need to set a custom prefix for security purposes or topic segregation
+    #[instrument(level = "info", skip_all)]
+    pub async fn new(
+        store: S,
+        client: Client,
+        topic_prefix: Option<&str>,
+    ) -> anyhow::Result<Server<S>> {
+        // Trim off any spaces or trailing/preceding dots
+        let prefix = topic_prefix
+            .unwrap_or(DEFAULT_WADM_TOPIC_PREFIX)
+            .trim()
+            .trim_matches('.')
+            .to_owned();
+        if prefix.is_empty() {
+            anyhow::bail!("Given prefix was empty")
+        }
```

Comment:
double checking, in the default case you expect the previs to be `wadm.api`, with all whitespace and leading/trailing periods removed?

Reply:
Yes, that is correct